### DISCUSSION TITLE
Undo vs History feature

### DISCUSSION THREAD

**Discussion by mikekucera** [cytoscape/cytoscape-explore#119] [Ideas]
### UNDO feature

- Description
    - A Standard undo feature that is common in many apps. 
    - Add two buttons in the top toolbar for Undo and Redo.
    - Create a client side component that maintains an “Undo Stack”. The stack contains objects that represent “Undo Edits”. Each edit contains the information required to Undo/Redo a specific user operation. Undo edits are “posted” (ie pushed) onto the stack when a user completes a specific operation that modifies the network. 
    - Undo pops the top edit off the stack and runs its undo() method.
    - The Undo stack has a maximum depth (usually 10-20 edits).
    - This service would be transient. If the user closes their browser tab then the undo history is lost.
- Concerns
  - There are two ways this could be implemented: UI driven or Event driven.
    - Event Driven
        - A global event handler, that listens to pretty much all types of events, is responsible for posting undo edits. 
        - The challenge with this approach is that it’s very difficult to determine when a sequence of events corresponds to a single user action. This becomes even more difficult when the application is highly interactive. For example the user performs a clustered layout, then makes some adjustments with the sliders. How many undo edits should be posted? One for the layout and then another for the slider adjustment? Or just a single edit comprising all the changes made in the layout panel? And how will the event handler know when the user is done adjusting the layout?
        - If event timing changes then it could break the undo feature in unexpected ways.
    - UI Driven
        - The various React UI components are responsible for posting Undo edits.
        - This has the advantage that we can use UI event/state to determine when a “User action” is complete. It can be very precise and clear when Undo edits are posted.
        - The big disadvantage is that Undo becomes a crosscutting concern that has to be weaved into every UI component in the entire application. UI components will become more complex, because they will have to store extra state to determine when a user action starts and ends. 
        - Inevitably we will miss some user actions and they won't be undoable. This happens all the time in cytoscape desktop.
- How should this work with the synchronization system?
    - Becomes very awkward and complex. 
    - Should I be able to Undo something that another user performed?
    - Should there be a single global Undo stack for all instances of the network?
    - Difficult mental model for the user to understand.

### History Feature

- Description
    - Similar to the “version history” feature in google docs or the “local history” feature in eclipse.
    - The system automatically takes “snapshots” of the network at specific points in time (discussed below). A “history” or “version” panel is available that allows the user to see thumbnails of the snapshots with the date/time. The user can restore an old snapshot, which overwrites the contents of the network.
    - Snapshots are created automatically at specific time intervals, or triggered by certain events.
    - The snapshot history can be pruned to avoid accumulating too many snapshots. The pruning gets more aggressive for older snapshots. For example there could be 5 snapshots from the last hour, 3 snapshots from the last day, 3 snapshots from the last week, 1 snapshot from the last month, etc…
- Advantages (compared to the Undo feature)
    - Snapshots can be taken on the server from the master copy of the network. Snapshots are available across all tab instances of a network because they are persisted in couchDB on the server. There is no awkwardness having this feature alongside the synchronization features.
    - The user can access snapshots long into the future.
    - The History Service runs in the background, it does not require any changes in existing UI components.
- Concerns
    - Performance. 
        - How does storing multiple versions of a network (potentially a large network) affect the storage requirements of the CE server? 
        - How long will it take to create and save a snapshot?
    - Not fine-grained like undo.
        - The user will not be able to undo specific operations (like setting a single node to a different color). They will only be able to go back it time by like 5 minutes. It won’t be clear to the user exactly what operations they are undoing by going back by an arbitrary amount of time. This could be mitigated by taking a lot of snapshots.

![PhotoEd - History](https://user-images.githubusercontent.com/5350528/146240273-ed9f9486-bb0d-4eb8-9366-6f63a00a3754.png)

### Proposal: Implement the History Feature

I believe the History feature is superior because:
- Provides more value to the user.
- Implementation will not be tightly coupled to the UI like Undo would be.
- Is not transient.
- Straightforward mental model, no confusing corner cases like trying to synchronize undo actions.

High level overview:
- Create a service for generating and storing network thumbnails.
    - By making this its own standalone service it can be made available to other features that might also require thumbnails. 
    - How to generate the thumbnails? Should it be done on the server after a snapshot is taken? Should it be done in the browser?
    - How/where should the thumbnail be stored?
- Save snapshots of a network.
    - Have a background thread on the sever that periodically takes snapshots of a network.
    - This service is also responsible for pruning the network history.
        - How will this work with couchDB?
        - How will this scale?
- Create a History panel UI
    - Shows a scrollable list of network thumbnails with the date/time.
    - Actions to restore a snapshot.
    - Should we have the ability to view a snapshot? Probably in a read only state.
        - Another option would be to have a small thumbnail for the panel, then a larger thumbnail for a preview shown in a pop-up.

**Comment by maxkfranz** [comment]
Re. history:

- Snapshots can be stored separately from the main documents in order to not clutter up the document data.  It's also easier to purge all the history that way.
- Pruning of the history could occur when an operation happens (i.e. on write) rather than on a time schedule (e.g. a cron job).  Whenever a user makes an edit, check for whether history should be pruned for this doc.  This approach is less error-prone (e.g. coordinating cron jobs across multiple Node.js instances) and less expensive than the time schedule (e.g. a cron job could hold up the server 100% until it's finished).  It's generally cleaner, since the restrictions put on the history state are invariant with this approach, whereas with a cron-job approach you can be in-between periods.
- In any case, it would be beneficial to have the history code runnable on either the server or the client, rather than tied to one or the other.  Then we can use it wherever we like.
- Possibly: Snapshots can alternatively be generated on the client to save server CPU and memory expense.  The client would have to upload the images.
- Possibly:  Alternatively, the history could be stored only on the user's local machine. 
  - This forgoes cross-user and cross-computer undos.  If Joe goes from his Mac to his iPad, he can't access the history.
  - You don't need to worry about pruning as much -- or at all in the case where the history isn't persisted across reloads or tabs.
  - This may not be ideal, but it may be simpler to start with.

**Reply by maxkfranz** [reply]
Another one:  Users may want to create snapshots manually at certain moments in time, because snapshot intervals of, say, five minutes may not capture the desired state (e.g. the user was still in the middle of what they would consider a snapshot when a snapshot was taken).

Generally, it would be better if snapshots could be completely automatic.  So if we could use a good heuristic for when a user would tend to want a snapshot, then we'd have the same effect.

One simple heuristic would be to snapshot the end of a batch of edits, e.g. (where `x` represents an edit and `^` represents a snapshot):


```
--------------------------- Time --------------------------->

xxx      xx          xxxxxx         xxxxxxxx
  ^       ^               ^                ^
```

We also need rules for how history pruning works, e.g.:

- Keep the last N snapshots
- Keep all snapshots within the last T minutes
- Keep N most-recent snapshots and also one old snapshot (e.g. yesterday's graph)
- Etc.
